flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java

```diff
@@ -0,0 +1,697 @@
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.addVertexIndexPrefixInVertexName;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.connect;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.createAndInitializeJobGraph;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.createChain;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.createSourceChainInfo;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.isChainable;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.isSourceChainable;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.markSupportingConcurrentExecutionAttempts;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.preValidate;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.serializeOperatorCoordinatorsAndStreamConfig;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.setAllOperatorNonChainedOutputsConfigs;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.setManagedMemoryFraction;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.setPhysicalEdges;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.setSlotSharingAndCoLocation;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.setVertexDescription;
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.validateHybridShuffleExecuteInBatchMode;
+
+/** Default implementation for {@link AdaptiveGraphGenerator}. */
+@Internal
+public class AdaptiveGraphManager implements AdaptiveGraphGenerator {
+
+    private final StreamGraph streamGraph;
+
+    private final JobGraph jobGraph;
+
+    private final StreamGraphHasher defaultStreamGraphHasher;
+
+    private final List<StreamGraphHasher> legacyStreamGraphHasher;
+
+    private final Executor serializationExecutor;
+
+    private final AtomicInteger vertexIndexId;
+
+    private final StreamGraphContext streamGraphContext;
+
+    private final Map<Integer, byte[]> hashes;
+
+    private final List<Map<Integer, byte[]>> legacyHashes;
+
+    // Records the id of stream node which job vertex is created.
+    private final Map<Integer, Integer> frozenNodeToStartNodeMap;
+
+    // When the downstream vertex is not created, we need to cache the output.
+    private final Map<Integer, Map<StreamEdge, NonChainedOutput>> intermediateOutputsCaches;
+
+    // Records the id of the stream node that produces the IntermediateDataSet.
+    private final Map<IntermediateDataSetID, Integer> intermediateDataSetIdToProducerMap;
```

Comment:
To confirm, is it just used for testing? Or it will be used in production code in later PRs?

Reply:
> To confirm, is it just used for testing? Or it will be used in production code in later PRs?

Will be used in later PRs.